File: python/06_045.py
# Sort Colors (Dutch National Flag problem)
def sortColors(nums):
    low = curr = 0
    high = len(nums) - 1
    
    while curr <= high:
        if nums[curr] == 0:
            nums[low], nums[curr] = nums[curr], nums[low]
            low += 1
            curr += 1
        elif nums[curr] == 2:
            nums[curr], nums[high] = nums[high], nums[curr]
            high -= 1
        else:
            curr += 1

File: data structures and algorithms/cs/06_045.cs
/**
 * Dutch national flag algorithm for sorting colors
 */
class Solution
{
   /**
    * Sort array containing only 0s, 1s, and 2s
    */
   public void sortColors(int[] nums)
   {
      int low = 0;
      int curr = 0;
      int high = nums.length - 1;

      while (curr <= high)
      {
         if (nums[curr] == 0)
         {
            // Move 0 to the left section
            int temp = nums[low];
            nums[low] = nums[curr];
            nums[curr] = temp;
            low++;
            curr++;
         }
         else if (nums[curr] == 2)
         {
            // Move 2 to the right section
            // Don't increment curr as swapped element needs checking
            int temp = nums[high];
            nums[high] = nums[curr];
            nums[curr] = temp;
            high--;
         }
         else
         {
            // Element is 1, keep in middle section
            curr++;
         }
      }
   }
}